### PR TITLE
Update electorate versioned spike

### PR DESCRIPTION
**Pull request by alpe**
Spike for ORM framework level versioning and utilities:

* New bucket sub types for id generation and versioning
* New Index to reference latest versions to an ID
* Combined Key to reference an ID and version in protobuf: `VersionedIDRef`
* Example implementation for `Electorate` + `ElectionRules` creation/updates

not impl delete

### CODE REVIEW

**Comment by husio** [comment]
I like the typed ID. I think this alone could be the functionality.

I see this implementation provides cool functionality. I wonder if we need all of it right now and if instead, we could use a simplified and less powerful version.

File: orm/codec.proto [hunk 1]
```diff
@@ -11,3 +13,12 @@ message MultiRef {
 message Counter {
   int64 count = 1;
 }
+
+// VersionedID is the combination of document ID and version number.
+message VersionedIDRef {
```
Comment (husio) [inline orm/codec.proto]: Would it make sense to rename it to `VersionedID` and use if both as an object ID and a reference type, the same as single value ID is used?
A versioned bucket could then require a `GetVersionedID() VersionedID` method to be present, which is is generated by the protobuf compiler.

File: orm/bucket.go [hunk 2]
```diff
@@ -51,17 +51,17 @@ type Bucket struct {
 var _ weave.QueryHandler = Bucket{}
 
 type namedIndex struct {
-	Index
```
Comment (ethanfrey) [inline orm/bucket.go]: I dont really understand the changes here. 

Why and what is the goal?
For versions or general cleanup?

In any case, it does bring up the issue that we need to consider how versioned data interacts with secondary indexes. 

My gut feel is only to index latest. Older versions only available with direct (Id, version) reference

File: orm/idgen_bucket.go [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+}
+
+type IDGenBucket struct {
+	Bucket
+	idGen IDGenerator
+}
+
+func WithSeqIDGenerator(b Bucket, seqName string) IDGenBucket {
+	seq := b.Sequence(seqName)
+	return WithIDGenerator(b, IDGeneratorFunc(func(db weave.KVStore, _ CloneableData) []byte {
```
Comment (ethanfrey) [inline orm/idgen_bucket.go]: But this means the same sequence is used for all objects.

Eg, 
Id 1,version 1
Id 2,version 2
Id 1,version 3...

Dont we need to read from the actual object

File: orm/model.go [hunk 4]
```diff
@@ -0,0 +1,5 @@
+package orm
+
+func (m VersionedIDRef) NextVersion() VersionedIDRef {
+	return VersionedIDRef{ID: m.ID, Version: m.Version + 1}
```
Comment (ethanfrey) [inline orm/model.go]: Okay, so this is where the version updates. I could really use better comments or docs

File: orm/versioning_bucket.go [hunk 5]
```diff
@@ -0,0 +1,84 @@
+package orm
+
+import (
+	"github.com/iov-one/weave"
+	"github.com/iov-one/weave/errors"
+)
+
+const latestVersionIndexName = "latest"
+
+type VersioningBucket struct {
```
Comment (ethanfrey) [inline orm/versioning_bucket.go]: Okay, interesting idea here. And good use to capture all code. 

I see the complexities of maintaining full objects as multiple versions. Powerful, but maybe overly complex. I am happy to discuss a simpler alternative today.

**Comment by ruseinov** [comment]
Shall we close this as in my understanding @husio is implementing the outcome of this idea?

**Comment by ethanfrey** [comment]
This is actually a different story.

@husio is working on schema migrations. That is the schema of the data (add a field with defaults, change processing).

This is about multiple versions of one document (eg. latest).

We had a nice long talk in Berlin and I think Alex was excited about finishing this up. I made many comments on how to simplify this to just the core idea, and not affect orm so much.

Please leave this open and I will let Alex finish this. This is to allow modifying the electorate (for future elections) during a running election without affecting that one.

**Comment by ruseinov** [comment]
aha, I didn’t read in the code too much, so just assumed the versioning story came from this.

Got it! will take a closer look a bit later on to familiarize.

**Comment by alpe** [comment]
Closing this now as discussed. Will do a new one this week that should be easier to read.

**Pull request closed without merging**